File: lab1/insertionSort.c
#include <stdio.h>
#include <stdlib.h> // For rand and srand
#include <time.h>   // For measuring execution time

void insertionSort(int arr[], int n) {
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;

        // Compare and shift elements
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

int main() {
    int n = 1000; // Array size
    int arr[n];
    int originalArr[n];
    int iterations = 10000; // Number of iterations

    // Seed for random number generation
    srand(time(NULL));

    // Generate a random array
    for (int i = 0; i < n; i++) {
        originalArr[i] = rand() % 10000;
    }

    double total_time = 0.0;

    for (int iter = 0; iter < iterations; iter++) {
        // Copy the original array to ensure each sort is on the same data
        for (int i = 0; i < n; i++) {
            arr[i] = originalArr[i];
        }

        clock_t start = clock();
        insertionSort(arr, n);
        clock_t end = clock();

        total_time += ((double)(end - start)) / CLOCKS_PER_SEC;
    }

    double average_time = total_time / iterations;
    printf("Average time taken by Insertion Sort in C (over %d runs): %f seconds\n", iterations, average_time);

    return 0;
}

File: lab1/Insertion_sort.py
import random
import time

def insertion_sort(arr):
    steps = 0
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
            steps += 2  # Increment steps for comparison and assignment
        arr[j + 1] = key
        steps += 1  # Increment step for the final assignment
    return arr, steps

# Parameters
n = 1000  # Size of the array
iterations = 10000  # Number of runs

# Generate random array
original_arr = [random.randint(1, 10000) for _ in range(n)]

total_time = 0
total_steps = 0

for _ in range(iterations):
    # Make a copy of the original array for each iteration
    arr = original_arr[:]
    
    start = time.time()
    _, steps = insertion_sort(arr)
    end = time.time()
    
    total_time += (end - start)
    total_steps += steps

# Calculate average time and steps
average_time = total_time / iterations
average_steps = total_steps / iterations

print(f"Average time taken by Insertion Sort in Python over {iterations} runs: {average_time:.6f} seconds")
print(f"Average steps taken by Insertion Sort over {iterations} runs: {average_steps:.2f}")
